Create Association Rules for the Market Basket Analysis for the given Threshold

# Market Basket Analysis

In [1]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=90d4356bb8141a0ce7583c05fcaebc268af972699d1620ee657523f46845165b
  Stored in directory: /root/.cache/pip/wheels/77/3d/a6/d317a6fb32be58a602b1e8c6b5d6f31f79322da554cad2a5ea
Successfully built apyori


In [2]:
import pandas as pd
import numpy as np
from apyori import apriori

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
# store_data = pd.read_csv("/content/gdrive/MyDrive/Datasets/GroceryStoreDataSet.csv",encoding='latin-1')
store_data= pd.read_csv("/content/drive/MyDrive/Machine Learning /datasets/04_GroceryStoreDataSet .csv", names = ['transaction'], sep = ',')
store_data

transaction
0               MILK,BREAD,BISCUIT
1    BREAD,MILK,BISCUIT,CORNFLAKES
2              BREAD,TEA,BOURNVITA
3             JAM,MAGGI,BREAD,MILK
4                MAGGI,TEA,BISCUIT
5              BREAD,TEA,BOURNVITA
6             MAGGI,TEA,CORNFLAKES
7          MAGGI,BREAD,TEA,BISCUIT
8              JAM,MAGGI,BREAD,TEA
9                       BREAD,MILK
10  COFFEE,COCK,BISCUIT,CORNFLAKES
11  COFFEE,COCK,BISCUIT,CORNFLAKES
12          COFFEE,SUGER,BOURNVITA
13               BREAD,COFFEE,COCK
14             BREAD,SUGER,BISCUIT
15         COFFEE,SUGER,CORNFLAKES
16           BREAD,SUGER,BOURNVITA
17              BREAD,COFFEE,SUGER
18              BREAD,COFFEE,SUGER
19      TEA,MILK,COFFEE,CORNFLAKES

In [5]:
store_data.head()

transaction
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [6]:
store_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 1 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   transaction  20 non-null     object
dtypes: object(1)
memory usage: 292.0+ bytes


In [7]:
#Renaming the column to PRODUCTS

store_data.rename(columns = {'transaction':'PRODUCTS'}, inplace = True)
store_data

PRODUCTS
0               MILK,BREAD,BISCUIT
1    BREAD,MILK,BISCUIT,CORNFLAKES
2              BREAD,TEA,BOURNVITA
3             JAM,MAGGI,BREAD,MILK
4                MAGGI,TEA,BISCUIT
5              BREAD,TEA,BOURNVITA
6             MAGGI,TEA,CORNFLAKES
7          MAGGI,BREAD,TEA,BISCUIT
8              JAM,MAGGI,BREAD,TEA
9                       BREAD,MILK
10  COFFEE,COCK,BISCUIT,CORNFLAKES
11  COFFEE,COCK,BISCUIT,CORNFLAKES
12          COFFEE,SUGER,BOURNVITA
13               BREAD,COFFEE,COCK
14             BREAD,SUGER,BISCUIT
15         COFFEE,SUGER,CORNFLAKES
16           BREAD,SUGER,BOURNVITA
17              BREAD,COFFEE,SUGER
18              BREAD,COFFEE,SUGER
19      TEA,MILK,COFFEE,CORNFLAKES

In [8]:
# Creating a list of items in every transaction.
#The list will work as a training set from where we can generate the list of Association Rules.

list_items = list(store_data["PRODUCTS"].apply(lambda x:x.split(',')))
list_items

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [9]:
from mlxtend.preprocessing import TransactionEncoder

encode = TransactionEncoder()
encode_data = encode.fit(list_items).transform(list_items)
store_data = pd.DataFrame(encode_data,columns = encode.columns_)
store_data.head()


BISCUIT  BOURNVITA  BREAD   COCK  COFFEE  CORNFLAKES    JAM  MAGGI   MILK  \
0     True      False   True  False   False       False  False  False   True   
1     True      False   True  False   False        True  False  False   True   
2    False       True   True  False   False       False  False  False  False   
3    False      False   True  False   False       False   True   True   True   
4     True      False  False  False   False       False  False   True  False   

   SUGER    TEA  
0  False  False  
1  False  False  
2  False   True  
3  False  False  
4  False   True

In [10]:
#Applying apriori algorithm
#The lift value of an association rule is the ratio of the confidence of the rule and the expected confidence of the rule.

association_rules = apriori(list_items, min_support=0.0045, min_confidence=0.2, min_lift=3, min_length=2)
association_list = list(association_rules)


In [11]:
#Generating the association rule between the items

for i in range(0, len(association_list)):
    print(association_list[i][0])

frozenset({'MAGGI', 'JAM'})
frozenset({'BISCUIT', 'COFFEE', 'COCK'})
frozenset({'BISCUIT', 'CORNFLAKES', 'COCK'})
frozenset({'BISCUIT', 'COFFEE', 'CORNFLAKES'})
frozenset({'BISCUIT', 'MAGGI', 'TEA'})
frozenset({'COFFEE', 'BOURNVITA', 'SUGER'})
frozenset({'CORNFLAKES', 'BREAD', 'MILK'})
frozenset({'BREAD', 'MAGGI', 'JAM'})
frozenset({'CORNFLAKES', 'COFFEE', 'COCK'})
frozenset({'CORNFLAKES', 'COFFEE', 'MILK'})
frozenset({'CORNFLAKES', 'COFFEE', 'TEA'})
frozenset({'COFFEE', 'MILK', 'TEA'})
frozenset({'CORNFLAKES', 'MILK', 'TEA'})
frozenset({'MILK', 'MAGGI', 'JAM'})
frozenset({'MAGGI', 'JAM', 'TEA'})
frozenset({'BISCUIT', 'MILK', 'BREAD', 'CORNFLAKES'})
frozenset({'BISCUIT', 'BREAD', 'MAGGI', 'TEA'})
frozenset({'BISCUIT', 'COFFEE', 'CORNFLAKES', 'COCK'})
frozenset({'MILK', 'BREAD', 'MAGGI', 'JAM'})
frozenset({'BREAD', 'MAGGI', 'JAM', 'TEA'})
frozenset({'CORNFLAKES', 'COFFEE', 'MILK', 'TEA'})


In [12]:
#####  APRIORI 2

In [13]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules
df = pd.read_csv("/content/drive/MyDrive/Machine Learning /datasets/04_GroceryStoreDataSet .csv", names = ['transaction'], sep = ',')

pandas for dataframe management
mlxtend (machine learning extensions) is a Python library of useful tools for day-to-day data science tasks.

mlxtend.preprocessing for data preprocessing before performing the task.

mlxtend.frequent_patterns for association rule task.

load data as a variable df

In [14]:
#2. One-hot encoding
#We have to change the data to the appropriate format before inputting it into the Apriori algorithm. The goal is a one-hot DataFrame.

#Change the DataFrame to a list of lists.

df = list(df["transaction"].apply(lambda x:x.split(",")))


In [15]:
#df = list(df["transaction"].apply(lambda x:x.split(",")))

In [16]:
#Use TransactionEncoder from the mlxtend library to change the list of transactions to a one-hot array.

In [17]:
one_hot_transformer = TransactionEncoder()
df_transform = one_hot_transformer.fit_transform(df)

In [18]:
#Change the one-hot array to a DataFrame.
df = pd.DataFrame(df_transform,columns=one_hot_transformer.columns_)

This is our desired format, one-hot. The columns are items in the store and each row represents a transaction. If the value is True, that item is sold in that transaction. Now, the data is ready to be fed to the algorithm

In [19]:
#3. Find the frequent itemsets using Apriori

df = apriori(df, min_support = 0.2, use_colnames = True)
df.sort_values(['support'],ascending=False, inplace = True)

 the minimum support to 0.2, so it will filter only a set of item(s) whose support is greater than 0.2 (or more than 20% from 19 transactions).

Only these itemsets which are considered important will proceed to the association rule.

In [20]:
#4. Association rule

df_ar = association_rules(df, metric="lift", min_threshold=1)
df_ar

antecedents   consequents  antecedent support  consequent support  \
0       (BREAD)        (MILK)                0.65                0.25   
1        (MILK)       (BREAD)                0.25                0.65   
2       (BREAD)       (SUGER)                0.65                0.30   
3       (SUGER)       (BREAD)                0.30                0.65   
4  (CORNFLAKES)      (COFFEE)                0.30                0.40   
5      (COFFEE)  (CORNFLAKES)                0.40                0.30   
6      (COFFEE)       (SUGER)                0.40                0.30   
7       (SUGER)      (COFFEE)                0.30                0.40   
8       (MAGGI)         (TEA)                0.25                0.35   
9         (TEA)       (MAGGI)                0.35                0.25   

   support  confidence      lift  representativity  leverage  conviction  \
0      0.2    0.307692  1.230769               1.0    0.0375    1.083333   
1      0.2    0.800000  1.230769               1.0    0.0375    1.750000   
2      0.2    0.307692  1.025641               1.0    0.0050    1.011111   
3      0.2    0.666667  1.025641               1.0    0.0050    1.050000   
4      0.2    0.666667  1.666667               1.0    0.0800    1.800000   
5      0.2    0.500000  1.666667               1.0    0.0800    1.400000   
6      0.2    0.500000  1.666667               1.0    0.0800    1.400000   
7      0.2    0.666667  1.666667               1.0    0.0800    1.800000   
8      0.2    0.800000  2.285714               1.0    0.1125    3.250000   
9      0.2    0.571429  2.285714               1.0    0.1125    1.750000   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.535714  0.285714   0.076923    0.553846  
1       0.250000  0.285714   0.428571    0.553846  
2       0.071429  0.266667   0.010989    0.487179  
3       0.035714  0.266667   0.047619    0.487179  
4       0.571429  0.400000   0.444444    0.583333  
5       0.666667  0.400000   0.285714    0.583333  
6       0.666667  0.400000   0.285714    0.583333  
7       0.571429  0.400000   0.444444    0.583333  
8       0.750000  0.500000   0.692308    0.685714  
9       0.865385  0.500000   0.428571    0.685714

The association_rules function will automatically calculate key metrics of our transaction data including support, confidence, lift, leverage, and conviction.

The result is required interpretation for further business actions.

Interpretation examples

From the highest confidence in index 0, the confidence is 0.8. It means that the customer who buys milk will buy bread for 80%. However, you have to keep in mind that confidence is not everything. The high confidence in this row is due to the high support of bread (consequence support 0.65) which means bread occurs in many transactions, so it will not make a business impact if we try to sell bread with milk anyway.
The better metric is lift or leverage. Index 8 has the highest lift and leverage at 2.28 and 0.1125 respectively. It means that the customers who buy tea are likely to buy Maggi as well.
What we can do after knowing this?

Change shelf layout to sell Maggi close to tea.
Promoting Maggi together with tea.